example/project_benchmark.exs

```diff
@@ -0,0 +1,85 @@
+defmodule Clickhouse.Benchmark.Performance do
+  use Benchee.Benchmark
+
+  # Since it's a module, any other constructs may be used,
+  # such as defs, defmacro, etc.
+  def map_fun(i) do
+    [i]
+  end
+
+  # Global benchmark setup. Optional.
+  #
+  # Called before any benchmark in the module.
+  #
+  # If defined, must return {:ok, state} for benchmarks to run
+  setup do
+    {:ok, nil}
+  end
+
+  # Global benchmark teardown. Optional.
+  #
+  # Called after all benchmarks have finished
+  # (and their possible local teardowns had been called).
+  #
+  # Can return anything.
+  teardown do
+    :anything
+  end
+
+  # Benchmarks. Module may have many of them.
+  benchmark "Flattening list from map", # Name. Mandatory.
+  warmup: 0, time: 1          # Opts (see Benchee.run). Optional.
+    do                                  # Do block. Mandatory.
+
+    # Benchmark setup. Optional
+    #
+    # If global setup is defined,
+    # implicit variable "state" is bound to it's result
+    #
+    # If defined, must return {:ok, state} for benchmarks to run
+    setup do
+      {:ok, fn x -> [x] end}
+    end
+
+    # Benchmark teardown. Optional
+    #
+    # Implicit variable: state, as it is returned from local setup
+    # (or from global setup, if no local setup is defined)
+    #
+    # Can return anything
+    teardown do
+      :anything
+    end
+
+    # Inputs: the same as passing inputs via :input option
+    #
+    # Accepts either an expression or a do block as a 2nd argument
+    input "Small", Enum.to_list(1..100)
+    input "Medium" do
+      n = 10_000
+      Enum.to_list(1..n)
+    end
+    input "Bigger", Enum.to_list(1..100_000)
+
+
+    # Benchmark scenarios
+    #
+    # Each scenario has an implicit variables:
+    # - state: state returned from local or global setup
+    # And if any input is passed (either with option or as input directive):
+    # - input: data for benchmark
```

Comment:
Honestly I did grit my teeth when writing this part. Although I somewhat despise magic forced implicits and implicit variables, I must admit that with implicits benchmark code looks cleaner.

Nonetheless, if you can give me examples of parameters/variables that we might be adding in the future, I will happily change implicits to map.

Reply:
With the map I just meant how you can reuse what setup returned in the tests: https://hexdocs.pm/ex_unit/ExUnit.Callbacks.html#setup_all/1-examples

Note sure what we'd need the examples for, overall, this would give people the possibility to put in there whatever data they need.

Not sure if that's the best. In "normal" benchee it's solved through just giving the function an argument where those values can be accomulated was the solution. Not certain one could do to give `do` blcoks an argument but I think I'd prefer it over implictness :)